### PR TITLE
react: Export type representing second argument of ForwardRefRenderFunction

### PR DESCRIPTION
**Pull request by wereHamster**
The type is not the same as `Ref`, this one allows `MutableRefObject` instead of the readonly `RefObject`.

Not sure why the original type didn't use `RefCallback`. Is there any particular reason?

# Motivation

```jsx
const Root = "div";

interface Props extends React.ComponentPropsWithoutRef<typeof Root> {
}

function Component(props: Props, ref: ???) {
  return <Root ref={ref} {...props} />;
}

export default React.forwardRef(Component);
```

There is currently no easy way to express the type of the `ref` function argument, `React.Ref<React.ElementRef<typeof Root>>` isn't correct. After this change I can use `React.ForwardedRef<React.ElementRef<typeof Root>>`.

---

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test YOUR_PACKAGE_NAME`.)
- [ ] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [ ] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ ] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

### CODE REVIEW

**Review by eps1lon — CHANGES_REQUESTED** [review]
Please add a test showcasing what this change enables. If it restricts unsound types  then you can add `$ExpectError` where it would previously wouldn't error. If it allows types that were previously rejected you don't have to add any special annotations: just include the code in the corresponding test files.

File: types/react/test/index.ts [hunk 1]
```diff
@@ -408,10 +408,16 @@ interface ForwardingRefComponentProps {
     foo: number;
 }
 
-const ForwardingRefComponent = React.forwardRef((props: ForwardingRefComponentProps, ref: React.Ref<RefComponent>) => {
+const ForwardingRefComponent = React.forwardRef((props: ForwardingRefComponentProps, ref: React.ForwardedRef<RefComponent>) => {
```
Comment (eps1lon) [inline types/react/test/index.ts:411]: Was this change required (if not we should revert it to not cause confusion)? Just to make sure we're not causing unnecessary churn.

**Review by eps1lon — CHANGES_REQUESTED** [review]
Looks good, one question remains.

File: types/react/test/index.ts (hunk 1 above)
Comment (wereHamster) [reply types/react/test/index.ts]: Not strictly necessary. I removed it.

**Review by eps1lon — APPROVED** [review]

**Comment by wereHamster** [comment]
Ready to merge
